README.md
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3660936.svg)](https://doi.org/10.5281/zenodo.3660936)

# EigenCuda

Offload the [Eigen3](http://eigen.tuxfamily.org/index.php?title=Main_Page) matrix matrix multiplication to an Nvidia GPU
using [cublas](https://docs.nvidia.com/cuda/cublas/index.html).

## CMake Installation

To compile execute:
```
cmake -H. -Bbuild && cmake --build build
```

To Debug compile as:
```
cmake -H. -Bbuild  -DCMAKE_BUILD_TYPE=Debug && cmake --build build
```

## Dependencies

This packages assumes that you have installed the following packages:
  
  * [Cudatoolkit](https://anaconda.org/anaconda/cudatoolkit)
  * [Eigen3](http://eigen.tuxfamily.org/index.php?title=Main_Page)

## Usage
### Matrix Multiplication
```cpp
#include "eigencuda.hpp"
#include "cudapipeline.hpp"

using eigencuda::CudaPipeline;
using eigencuda::CudaMatrix;

  // Call the class to handle GPU resources
  CudaPipeline cuda_pip;

Eigen::MatrixXd A = Eigen::MatrixXd::Zero(2, 2);
Eigen::MatrixXd B = Eigen::MatrixXd::Zero(3, 2);
Eigen::MatrixXd C = Eigen::MatrixXd::Zero(3, 2);
Eigen::MatrixXd D = Eigen::MatrixXd::Zero(3, 2);
Eigen::MatrixXd X = Eigen::MatrixXd::Zero(3, 2);
Eigen::MatrixXd Y = Eigen::MatrixXd::Zero(3, 2);
Eigen::MatrixXd Z = Eigen::MatrixXd::Zero(3, 2);

// Define matrices
A << 1., 2., 3., 4.;
B << 5., 6., 7., 8., 9., 10.;
C << 9., 10., 11., 12., 13., 14.;
D << 13., 14., 15., 16., 17., 18.;
X << 23., 34., 31., 46., 39., 58.;
Y << 39., 58., 47., 70., 55., 82.;
Z << 55., 82., 63., 94., 71., 106.;

std::vector<Eigen::MatrixXd> tensor{B, C, D};
std::vector<Eigen::MatrixXd> results(3, Eigen::MatrixXd::Zero(3, 2));
CudaMatrix cuma_A{A, cuda_pip.get_stream()};
CudaMatrix cuma_B{3, 2, cuda_pip.get_stream()};
CudaMatrix cuma_C{3, 2, cuda_pip.get_stream()};

for (Index i = 0; i < 3; i++) {
  cuma_B.copy_to_gpu(tensor[i]);
  cuda_pip.gemm(cuma_B, cuma_A, cuma_C);
  results[i] = cuma_C;
}
// Expected results
bool pred_1 = X.isApprox(results[0]);
bool pred_2 = Y.isApprox(results[1]);
bool pred_3 = Z.isApprox(results[2]);
}
```


CHANGELOG.md
# Change log

# [0.4.0] 10/02/2020
### Changed
  - Split the memory management (`CudaMatrix`) from the [CUBLAS](https://docs.nvidia.com/cuda/cublas/index.html) invocation (`CudaPipeline`)
  - Moved all the allocation to the smart pointers inside `CudaMatrix`
 - Removed unused headers

# [0.3.0] 26/09/2019
### Added
 - Smart pointers to handle cuda resources
 - New CudaMatrix class
 - Use Eigen::MatrixXd
 - Check available memory in the GPU before computing

### Removed
 - Template class, implementation only for double available
 - Triple tensor product
 - Shapes struct


# [0.2.0] 27/08/2019
### Added
 - Tensor matrix multiplacation using [gemmbatched](https://docs.nvidia.com/cuda/CUBLAS/index.html#CUBLAS-lt-t-gt-gemmbatched).
 - [Async calls](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79) to memory copies.
 - Properly free memory after the tensor operation is done.

# [0.1.0]

### New
 - Use a template function to perform matrix matrix multiplacation using [CUBLAS](https://docs.nvidia.com/cuda/CUBLAS/index.html).
 - Use either *pinned* (**default**) or *pageable* memory, see [cuda optimizations](https://devblogs.nvidia.com/how-optimize-data-transfers-cuda-cc/).


src/benchmark.cc
#include "eigencuda.hpp"
#include <chrono>
#include <fstream>
#include <iostream>
#include <vector>

using eigencuda::Mat;

void write_vector(std::vector<int> sizes,
                  std::vector<std::tuple<double, double>> vs) {
  std::ofstream file;
  double x, y;

  file.open("times_benchmark.txt");
  for (unsigned i = 0; i < vs.size(); i++) {
    std::tie(x, y) = vs[i];
    file << sizes[i] << " " << x << " " << y << "\n";
  }
  file.close();
}

template <typename T>
std::tuple<double, double> benchmark_right_matrix_tensor(
    Mat<T> &A, std::vector<Mat<T>> tensor) {
  // chrono
  std::chrono::time_point<std::chrono::system_clock> start, end;
  std::chrono::duration<double> elapsed_time;

  // Cuda
  start = std::chrono::system_clock::now();
  eigencuda::EigenCuda<double> EC;
  std::vector<Mat<double>> rs = EC.right_matrix_tensor(A, tensor);
  end = std::chrono::system_clock::now();
  elapsed_time = end - start;
  auto gpu_time = elapsed_time.count();
  std::cout << "GPU tensor matrix product: " << gpu_time << " secs\n";

  // Call Eigen
  start = std::chrono::system_clock::now();
  std::vector<Mat<double>> qs;
  for (const auto &x : tensor) {
    Mat<double> r = x * A;
    qs.push_back(r);
  }
  end = std::chrono::system_clock::now();
  elapsed_time = end - start;
  auto cpu_time = elapsed_time.count();
  std::cout << "CPU tensor tensor product: " << cpu_time << " secs\n";

  return std::make_tuple(gpu_time, cpu_time);
}

void run_benchmark(std::vector<int> vs, bool pinned = false) {

  std::vector<std::tuple<double, double>> times;

  for (auto size : vs) {
    // Create CPU matrices
    std::cout << "running benchmark!\n";

    Mat<double> B = Mat<double>::Random(size, size + 20);

    std::string msg =
        (pinned) ? "Pinned Data Transfer" : "Pageable Data Transfer";
    std::cout << "size: " << size << "\n";
    std::cout << msg << "\n";

    // Benchmark for tensor product
    std::vector<Mat<double>> tensor;
    for (auto i = 0; i < 10; i++) {
      tensor.push_back(Mat<double>::Random(size + 20, size));
    }
    std::cout << "Running right matrix tensor benchmark\n";
    times.push_back(benchmark_right_matrix_tensor(B, tensor));
  }
  write_vector(vs, times);
}

void right_matrix_tensor() {
  // Define matrices and class to handle GPU resources
  eigencuda::EigenCuda<double> EC;

  // Call matrix multiplication GPU
  Mat<double> A = Mat<double>::Zero(2, 2);
  Mat<double> B = Mat<double>::Zero(3, 2);
  Mat<double> C = Mat<double>::Zero(3, 2);
  Mat<double> D = Mat<double>::Zero(3, 2);

  // Define matrices
  A << 1., 2., 3., 4.;
  B << 5., 6., 7., 8., 9., 10.;
  C << 9., 10., 11., 12., 13., 14.;
  D << 13., 14., 15., 16., 17., 18.;

  std::vector<Mat<double>> tensor{B, C, D};
  std::vector<Mat<double>> rs = EC.right_matrix_tensor(A, tensor);

  for (auto &x : rs) {
    std::cout << "result sum: " << x << "\n";
  }

  std::cout << "right matrix product succeeded!\n";
}

void triple_product() {
  // Define matrices and class to handle GPU resources
  eigencuda::EigenCuda<double> EC;
  // Call matrix multiplication GPU
  Mat<double> A = Mat<double>::Zero(2, 3);
  Mat<double> B = Mat<double>::Zero(3, 2);
  Mat<double> C = Mat<double>::Zero(3, 3);
  Mat<double> D = Mat<double>::Zero(3, 3);

  // Define matrices
  A << 1., 2., 3., 4., 5., 6.;
  B << 5., 6., 7., 8., 9., 10.;
  C << 9., 10., 11., 12., 13., 14., 15., 16., 17.;
  D << 13., 14., 15., 16., 17., 18., 19., 20., 21.;

  std::vector<Mat<double>> tensor{C, D};
  std::vector<Mat<double>> rs = EC.triple_tensor_product(A, B, tensor);

  for (auto &x : rs) {
    std::cout << "matrix: " << x << "\n";
  }

  // Check results
  assert(abs(rs[0].sum() - 12993.) < 1e-8);
  assert(abs(rs[1].sum() - 16773.) < 1e-8);

  std::cout << "triple product succeeded!\n";
}

int main() {

  bool pinned = false;
  std::vector<int> vs{100, 200, 500};

  // run_benchmark(vs, pinned);
  right_matrix_tensor();
  triple_product();
  return 0;
}
